src/main.rs

```diff
@@ -0,0 +1,150 @@
+use db::{Db, DbConnection, DbEvent};
+use futures::{future::FutureExt, stream::StreamExt};
+use futures_batch::ChunksTimeoutStreamExt;
+use serde::{Deserialize, Serialize};
+use std::{env, time::Duration};
+use structopt::StructOpt;
+use tokio_compat_02::FutureExt as FutureExt02;
+use tracing::*;
+
+mod db;
+mod postgres;
+
+#[derive(StructOpt, Debug, Serialize, Deserialize)]
+#[structopt(
+    about = "Insert events into a Postgres database.\nEnvironment variables:\n  PGPASSWORD: PostgreSQL password, if not provided with --password/-p\n  AX_EVENT_SERVICE_URI: URL to connect to the Event Service. Default: http://localhost:4454/api/"
+)]
+struct Opt {
+    #[structopt(long, short)]
+    from_start: bool,
+    #[structopt(
+        long,
+        short = "w",
+        about = "If not provided, read from the PGPASSWORD environment variable"
+    )]
+    password: Option<String>,
+    #[structopt(long, short, default_value = "1024")]
+    max_batch_size: usize,
+    #[structopt(long, short)]
+    db_name: String,
+    #[structopt(long, short, default_value = "localhost")]
+    host: String,
+    #[structopt(long, short, default_value = "5432")]
+    port: u16,
+    #[structopt(long, short)]
+    username: String,
+    #[structopt(long, short, default_value = "events")]
+    table: String,
+    #[structopt(about = "Subscription set to listen to", default_value = "[{}]")]
+    subscriptions: String,
+}
+
+pub fn main() -> Result<()> {
+    env_logger::init();
+    init_panic_hook();
+    let opt: Opt = Opt::from_args();
+
+    let password = opt
+        .password
+        .unwrap_or_else(|| env::var("PGPASSWORD").expect("Postgres password must be provided"));
+
+    let subscriptions: Vec<Subscription> =
+        serde_json::from_str(opt.subscriptions.as_str()).unwrap();
+    info!("Subscribing to: {:?}", subscriptions);
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(2)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let pg = Postgres::new(
+        opt.host,
+        opt.port,
+        opt.username,
+        password,
+        opt.db_name,
+        opt.table,
+    );
+
+    runtime.block_on(run_pipeline(
+        Box::new(pg),
+        subscriptions,
+        opt.max_batch_size,
+    ))
+}
+
+async fn run_pipeline<C: DbConnection + 'static>(
+    db: Box<dyn Db<C>>,
+    subscriptions: Vec<Subscription>,
+    max_batch_size: usize,
+) -> Result<()> {
+    let event_service = EventService::default();
+    debug!("Connected to EventService");
+
+    let present = event_service.get_offsets().compat().await?;
+    debug!("Offset map from swarm: {:?}", present);
+
+    let db = db.connect().await?;
+    let offsets = db.get_offsets().await?;
+    debug!("Offset map from database: {:?}", offsets);
+
+    info!(
+        "Database has {} events. Swarm has {} events. Getting {} events to get to the present",
+        &offsets.size(),
+        &present.size(),
+        &present - &offsets
```

Comment:
Just curious: That results in an offset map, right? Could you provide a sample print?

Reply:
looking at the source, it results in `&presents.size() - &offsets.size()`

Reply:
I've removed it anyway, since it does not take the subscriptions into account.